<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Splat simulation parameters • Splatter</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Splat simulation parameters">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-52309538-4"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-52309538-4');
</script>
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Splatter</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.9.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/splatter.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/splat_params.html">Splat simulation parameters</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Oshlack/splatter">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Splat simulation parameters</h1>
            
            <h4 class="date">Last updated: 18 April 2019</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/Oshlack/splatter/blob/master/vignettes/splat_params.Rmd"><code>vignettes/splat_params.Rmd</code></a></small>
      <div class="hidden name"><code>splat_params.Rmd</code></div>

    </div>

    
    
<p>This vignette describes the Splat simulation model and the parameters it uses in more detail.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(<span class="st">"splatter"</span>)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="co">#&gt; Loading required package: SingleCellExperiment</span></a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="co">#&gt; Loading required package: SummarizedExperiment</span></a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="co">#&gt; Loading required package: GenomicRanges</span></a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="co">#&gt; Loading required package: stats4</span></a>
<a class="sourceLine" id="cb1-6" data-line-number="6"><span class="co">#&gt; Loading required package: BiocGenerics</span></a>
<a class="sourceLine" id="cb1-7" data-line-number="7"><span class="co">#&gt; Loading required package: parallel</span></a>
<a class="sourceLine" id="cb1-8" data-line-number="8"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-9" data-line-number="9"><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></a>
<a class="sourceLine" id="cb1-10" data-line-number="10"><span class="co">#&gt; The following objects are masked from 'package:parallel':</span></a>
<a class="sourceLine" id="cb1-11" data-line-number="11"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-12" data-line-number="12"><span class="co">#&gt;     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,</span></a>
<a class="sourceLine" id="cb1-13" data-line-number="13"><span class="co">#&gt;     clusterExport, clusterMap, parApply, parCapply, parLapply,</span></a>
<a class="sourceLine" id="cb1-14" data-line-number="14"><span class="co">#&gt;     parLapplyLB, parRapply, parSapply, parSapplyLB</span></a>
<a class="sourceLine" id="cb1-15" data-line-number="15"><span class="co">#&gt; The following objects are masked from 'package:stats':</span></a>
<a class="sourceLine" id="cb1-16" data-line-number="16"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-17" data-line-number="17"><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></a>
<a class="sourceLine" id="cb1-18" data-line-number="18"><span class="co">#&gt; The following objects are masked from 'package:base':</span></a>
<a class="sourceLine" id="cb1-19" data-line-number="19"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-20" data-line-number="20"><span class="co">#&gt;     anyDuplicated, append, as.data.frame, basename, cbind,</span></a>
<a class="sourceLine" id="cb1-21" data-line-number="21"><span class="co">#&gt;     colnames, dirname, do.call, duplicated, eval, evalq, Filter,</span></a>
<a class="sourceLine" id="cb1-22" data-line-number="22"><span class="co">#&gt;     Find, get, grep, grepl, intersect, is.unsorted, lapply, Map,</span></a>
<a class="sourceLine" id="cb1-23" data-line-number="23"><span class="co">#&gt;     mapply, match, mget, order, paste, pmax, pmax.int, pmin,</span></a>
<a class="sourceLine" id="cb1-24" data-line-number="24"><span class="co">#&gt;     pmin.int, Position, rank, rbind, Reduce, rownames, sapply,</span></a>
<a class="sourceLine" id="cb1-25" data-line-number="25"><span class="co">#&gt;     setdiff, sort, table, tapply, union, unique, unsplit, which,</span></a>
<a class="sourceLine" id="cb1-26" data-line-number="26"><span class="co">#&gt;     which.max, which.min</span></a>
<a class="sourceLine" id="cb1-27" data-line-number="27"><span class="co">#&gt; Loading required package: S4Vectors</span></a>
<a class="sourceLine" id="cb1-28" data-line-number="28"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-29" data-line-number="29"><span class="co">#&gt; Attaching package: 'S4Vectors'</span></a>
<a class="sourceLine" id="cb1-30" data-line-number="30"><span class="co">#&gt; The following object is masked from 'package:base':</span></a>
<a class="sourceLine" id="cb1-31" data-line-number="31"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-32" data-line-number="32"><span class="co">#&gt;     expand.grid</span></a>
<a class="sourceLine" id="cb1-33" data-line-number="33"><span class="co">#&gt; Loading required package: IRanges</span></a>
<a class="sourceLine" id="cb1-34" data-line-number="34"><span class="co">#&gt; Loading required package: GenomeInfoDb</span></a>
<a class="sourceLine" id="cb1-35" data-line-number="35"><span class="co">#&gt; Loading required package: Biobase</span></a>
<a class="sourceLine" id="cb1-36" data-line-number="36"><span class="co">#&gt; Welcome to Bioconductor</span></a>
<a class="sourceLine" id="cb1-37" data-line-number="37"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-38" data-line-number="38"><span class="co">#&gt;     Vignettes contain introductory material; view with</span></a>
<a class="sourceLine" id="cb1-39" data-line-number="39"><span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span></a>
<a class="sourceLine" id="cb1-40" data-line-number="40"><span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></a>
<a class="sourceLine" id="cb1-41" data-line-number="41"><span class="co">#&gt; Loading required package: DelayedArray</span></a>
<a class="sourceLine" id="cb1-42" data-line-number="42"><span class="co">#&gt; Loading required package: matrixStats</span></a>
<a class="sourceLine" id="cb1-43" data-line-number="43"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-44" data-line-number="44"><span class="co">#&gt; Attaching package: 'matrixStats'</span></a>
<a class="sourceLine" id="cb1-45" data-line-number="45"><span class="co">#&gt; The following objects are masked from 'package:Biobase':</span></a>
<a class="sourceLine" id="cb1-46" data-line-number="46"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-47" data-line-number="47"><span class="co">#&gt;     anyMissing, rowMedians</span></a>
<a class="sourceLine" id="cb1-48" data-line-number="48"><span class="co">#&gt; Loading required package: BiocParallel</span></a>
<a class="sourceLine" id="cb1-49" data-line-number="49"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-50" data-line-number="50"><span class="co">#&gt; Attaching package: 'DelayedArray'</span></a>
<a class="sourceLine" id="cb1-51" data-line-number="51"><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></a>
<a class="sourceLine" id="cb1-52" data-line-number="52"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-53" data-line-number="53"><span class="co">#&gt;     colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges</span></a>
<a class="sourceLine" id="cb1-54" data-line-number="54"><span class="co">#&gt; The following objects are masked from 'package:base':</span></a>
<a class="sourceLine" id="cb1-55" data-line-number="55"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-56" data-line-number="56"><span class="co">#&gt;     aperm, apply, rowsum</span></a>
<a class="sourceLine" id="cb1-57" data-line-number="57"><span class="co">#&gt; Registered S3 methods overwritten by 'ggplot2':</span></a>
<a class="sourceLine" id="cb1-58" data-line-number="58"><span class="co">#&gt;   method         from </span></a>
<a class="sourceLine" id="cb1-59" data-line-number="59"><span class="co">#&gt;   [.quosures     rlang</span></a>
<a class="sourceLine" id="cb1-60" data-line-number="60"><span class="co">#&gt;   c.quosures     rlang</span></a>
<a class="sourceLine" id="cb1-61" data-line-number="61"><span class="co">#&gt;   print.quosures rlang</span></a>
<a class="sourceLine" id="cb1-62" data-line-number="62"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(<span class="st">"scater"</span>)</a>
<a class="sourceLine" id="cb1-63" data-line-number="63"><span class="co">#&gt; Loading required package: ggplot2</span></a>
<a class="sourceLine" id="cb1-64" data-line-number="64"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(<span class="st">"ggplot2"</span>)</a></code></pre></div>
<div id="the-base-splat-model" class="section level1">
<h1 class="hasAnchor">
<a href="#the-base-splat-model" class="anchor"></a>The base Splat model</h1>
<p>This figure, taken from the Splatter publication, describes the core of the Splat simulation model.</p>
<div class="figure">
<img src="splat-model.png" alt="Splat simulation model"><p class="caption">Splat simulation model</p>
</div>
<p>The Splat simulation uses a hierarchical probabilistic where different aspects of a dataset are generated from appropriate statistical distributions. The first stage generates a mean expression level for each gene. These are originally chosen from a Gamma distribution. For some genes that are selected to be outliers with high expression a factor is generated from a log-normal distribution. These factors are then multiplied by the median gene mean to create new means for those genes.</p>
<p>The next stage incorporates variation in the counts per cell. An expected library size (total counts) is chosen for each cell from a log-normal distribution. The library sizes are then used to scale the gene means for each cell, resulting in a range a counts per cell in the simulated dataset. The gene means are then further adjusted to enforce a relationship between the mean expression level and the variability.</p>
<p>The final cell by gene matrix of gene means is then used to generate a count matrix using a Poisson distribution. The result is a synthetic dataset consisting of counts from a Gamma-Poisson (or negative-binomial) distribution. An additional optional step can be used to replicate a “dropout” effect. A probability of dropout is generated using a logistic function based on the underlying mean expression level. A Bernoulli distribution is then used to create a dropout matrix which sets some of the generated counts to zero.</p>
<p>The model described here will generate a single population of cells but the Splat simulation has been designed to be as flexible as possible and can create scenarios including multiple groups of cells (cell types), continuous paths between cell types and multiple experimental batches. The parameters used to create these types of simulations and how they interact with the model are described below.</p>
</div>
<div id="splat-simulation-parameters" class="section level1">
<h1 class="hasAnchor">
<a href="#splat-simulation-parameters" class="anchor"></a>Splat simulation parameters</h1>
<p>Within Splatter the parameters for the Splat simulation model are held in the <code>SplatParams</code> object. Let’s create one of these objects and see what it looks like.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1">params &lt;-<span class="st"> </span><span class="kw"><a href="../reference/newParams.html">newSplatParams</a></span>()</a>
<a class="sourceLine" id="cb2-2" data-line-number="2">params</a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="co">#&gt; A Params object of class SplatParams </span></a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="co">#&gt; Parameters can be (estimable) or [not estimable], 'Default' or  'NOT DEFAULT' </span></a>
<a class="sourceLine" id="cb2-5" data-line-number="5"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb2-6" data-line-number="6"><span class="co">#&gt; Global: </span></a>
<a class="sourceLine" id="cb2-7" data-line-number="7"><span class="co">#&gt; (Genes)  (Cells)   [Seed] </span></a>
<a class="sourceLine" id="cb2-8" data-line-number="8"><span class="co">#&gt;   10000      100   243223 </span></a>
<a class="sourceLine" id="cb2-9" data-line-number="9"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb2-10" data-line-number="10"><span class="co">#&gt; 28 additional parameters </span></a>
<a class="sourceLine" id="cb2-11" data-line-number="11"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb2-12" data-line-number="12"><span class="co">#&gt; Batches: </span></a>
<a class="sourceLine" id="cb2-13" data-line-number="13"><span class="co">#&gt;     [Batches]  [Batch Cells]     [Location]        [Scale] </span></a>
<a class="sourceLine" id="cb2-14" data-line-number="14"><span class="co">#&gt;             1            100            0.1            0.1 </span></a>
<a class="sourceLine" id="cb2-15" data-line-number="15"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb2-16" data-line-number="16"><span class="co">#&gt; Mean: </span></a>
<a class="sourceLine" id="cb2-17" data-line-number="17"><span class="co">#&gt;  (Rate)  (Shape) </span></a>
<a class="sourceLine" id="cb2-18" data-line-number="18"><span class="co">#&gt;     0.3      0.6 </span></a>
<a class="sourceLine" id="cb2-19" data-line-number="19"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb2-20" data-line-number="20"><span class="co">#&gt; Library size: </span></a>
<a class="sourceLine" id="cb2-21" data-line-number="21"><span class="co">#&gt; (Location)     (Scale)      (Norm) </span></a>
<a class="sourceLine" id="cb2-22" data-line-number="22"><span class="co">#&gt;         11         0.2       FALSE </span></a>
<a class="sourceLine" id="cb2-23" data-line-number="23"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb2-24" data-line-number="24"><span class="co">#&gt; Exprs outliers: </span></a>
<a class="sourceLine" id="cb2-25" data-line-number="25"><span class="co">#&gt; (Probability)     (Location)        (Scale) </span></a>
<a class="sourceLine" id="cb2-26" data-line-number="26"><span class="co">#&gt;          0.05              4            0.5 </span></a>
<a class="sourceLine" id="cb2-27" data-line-number="27"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb2-28" data-line-number="28"><span class="co">#&gt; Groups: </span></a>
<a class="sourceLine" id="cb2-29" data-line-number="29"><span class="co">#&gt;      [Groups]  [Group Probs] </span></a>
<a class="sourceLine" id="cb2-30" data-line-number="30"><span class="co">#&gt;             1              1 </span></a>
<a class="sourceLine" id="cb2-31" data-line-number="31"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb2-32" data-line-number="32"><span class="co">#&gt; Diff expr: </span></a>
<a class="sourceLine" id="cb2-33" data-line-number="33"><span class="co">#&gt; [Probability]    [Down Prob]     [Location]        [Scale] </span></a>
<a class="sourceLine" id="cb2-34" data-line-number="34"><span class="co">#&gt;           0.1            0.5            0.1            0.4 </span></a>
<a class="sourceLine" id="cb2-35" data-line-number="35"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb2-36" data-line-number="36"><span class="co">#&gt; BCV: </span></a>
<a class="sourceLine" id="cb2-37" data-line-number="37"><span class="co">#&gt; (Common Disp)          (DoF) </span></a>
<a class="sourceLine" id="cb2-38" data-line-number="38"><span class="co">#&gt;           0.1             60 </span></a>
<a class="sourceLine" id="cb2-39" data-line-number="39"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb2-40" data-line-number="40"><span class="co">#&gt; Dropout: </span></a>
<a class="sourceLine" id="cb2-41" data-line-number="41"><span class="co">#&gt;     [Type]  (Midpoint)     (Shape) </span></a>
<a class="sourceLine" id="cb2-42" data-line-number="42"><span class="co">#&gt;       none           0          -1 </span></a>
<a class="sourceLine" id="cb2-43" data-line-number="43"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb2-44" data-line-number="44"><span class="co">#&gt; Paths: </span></a>
<a class="sourceLine" id="cb2-45" data-line-number="45"><span class="co">#&gt;         [From]         [Steps]          [Skew]    [Non-linear] </span></a>
<a class="sourceLine" id="cb2-46" data-line-number="46"><span class="co">#&gt;              0             100             0.5             0.1 </span></a>
<a class="sourceLine" id="cb2-47" data-line-number="47"><span class="co">#&gt; [Sigma Factor] </span></a>
<a class="sourceLine" id="cb2-48" data-line-number="48"><span class="co">#&gt;            0.8</span></a></code></pre></div>
<p>Like all the parameter objects in Splatter printing this object displays all the parameters required for this simulation. As we haven’t set any of the parameters the default values are shown but if we were to change any of them they would be highlighted. We can also see which parameters can be estimated by the Splat estimation procedure and which can’t. The default values have been chosen to be fairly realistic but it is recommended that estimation is used to get a simulation that is more like the data you are interested in. Parameters can be modified by setting them in the <code>SplatParams</code> object or by providing them directly to the simulation function.</p>
<p>The rest of this section provides details of all this parameters and explains how they can be used with examples.</p>
<div id="global-parameters" class="section level2">
<h2 class="hasAnchor">
<a href="#global-parameters" class="anchor"></a>Global parameters</h2>
<p>These parameters are used in every simulation model and control global features of the dataset produced.</p>
<div id="ngenes---number-of-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#ngenes---number-of-genes" class="anchor"></a><code>nGenes</code> - Number of genes</h3>
<p>The number of genes to simulate.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co"># Set the number of genes to 1000</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2">params &lt;-<span class="st"> </span><span class="kw"><a href="../reference/setParam.html">setParam</a></span>(params, <span class="st">"nGenes"</span>, <span class="dv">1000</span>)</a>
<a class="sourceLine" id="cb3-3" data-line-number="3"></a>
<a class="sourceLine" id="cb3-4" data-line-number="4">sim &lt;-<span class="st"> </span><span class="kw"><a href="../reference/splatSimulate.html">splatSimulate</a></span>(params, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb3-5" data-line-number="5"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/dim">dim</a></span>(sim)</a>
<a class="sourceLine" id="cb3-6" data-line-number="6"><span class="co">#&gt; [1] 1000  100</span></a></code></pre></div>
</div>
<div id="ncells---number-of-cells" class="section level3">
<h3 class="hasAnchor">
<a href="#ncells---number-of-cells" class="anchor"></a><code>nCells</code> - Number of cells</h3>
<p>The number of genes to simulate. In the Splat simulation this cannot be set directly but must be controlled using the <code>batchCells</code> parameter.</p>
</div>
<div id="seed---random-seed" class="section level3">
<h3 class="hasAnchor">
<a href="#seed---random-seed" class="anchor"></a><code>seed</code> - Random seed</h3>
<p>Seed to use for generating random numbers including selecting values from distributions. By changing this value multiple simulated datasets with the same parameters can be produced. Simulations produced using the same set of parameters and random seed should be identical but there may be differences between operating systems, software versions etc.</p>
</div>
</div>
<div id="batch-parameters" class="section level2">
<h2 class="hasAnchor">
<a href="#batch-parameters" class="anchor"></a>Batch parameters</h2>
<p>These parameters control experimental batches in the simulated dataset. The overall effect of how batch effects are included in the model is similar to technical replicates (i.e. the same biological sample sequenced multiple times). This means that the underlying structure is consistent between batches but a global technical signature is added may separate them.</p>
<div id="nbatches---number-of-batches" class="section level3">
<h3 class="hasAnchor">
<a href="#nbatches---number-of-batches" class="anchor"></a><code>nBatches</code> - Number of batches</h3>
<p>The number of batches in the simulation. This cannot be set directly but is controlled by setting <code>batchCells</code>.</p>
</div>
<div id="batchcells---cells-per-batch" class="section level3">
<h3 class="hasAnchor">
<a href="#batchcells---cells-per-batch" class="anchor"></a><code>batchCells</code> - Cells per batch</h3>
<p>A vector specifying the number of cells in each batch. The number of batches (<code>nBatches</code>) is equal to the length of the vector and the number of cells (<code>nCells</code>) is equal to the sum.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co"># Simulation with two batches of 100 cells</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">sim &lt;-<span class="st"> </span><span class="kw"><a href="../reference/splatSimulate.html">splatSimulate</a></span>(params, <span class="dt">batchCells =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">100</span>, <span class="dv">100</span>), <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb4-3" data-line-number="3"></a>
<a class="sourceLine" id="cb4-4" data-line-number="4"><span class="co"># PCA plot using scater</span></a>
<a class="sourceLine" id="cb4-5" data-line-number="5">sim &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/normalize">normalize</a></span>(sim)</a>
<a class="sourceLine" id="cb4-6" data-line-number="6"><span class="co">#&gt; Warning in .local(object, ...): using library sizes as size factors</span></a>
<a class="sourceLine" id="cb4-7" data-line-number="7"><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plot_reddim">plotPCA</a></span>(sim, <span class="dt">colour_by =</span> <span class="st">"Batch"</span>)</a></code></pre></div>
<p><img src="splat_params_files/figure-html/batches-1.png" width="700"></p>
</div>
<div id="batch-facloc---batch-factor-location-and-batch-facscale---batch-factor-scale" class="section level3">
<h3 class="hasAnchor">
<a href="#batch-facloc---batch-factor-location-and-batch-facscale---batch-factor-scale" class="anchor"></a><code>batch.facLoc</code> - Batch factor location and <code>batch.facScale</code> - Batch factor scale</h3>
<p>Batches are specified by generating a small scaling factor for each gene in each batch from a log-normal distribution. These factors are then applied to the underlying gene means in each batch. Modifying these parameters affects how different the batches are from each other by generating bigger or smaller factors.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="co"># Simulation with small batch effects</span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2">sim1 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/splatSimulate.html">splatSimulate</a></span>(params, <span class="dt">batchCells =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">100</span>, <span class="dv">100</span>),</a>
<a class="sourceLine" id="cb5-3" data-line-number="3">                      <span class="dt">batch.facLoc =</span> <span class="fl">0.001</span>, <span class="dt">batch.facScale =</span> <span class="fl">0.001</span>,</a>
<a class="sourceLine" id="cb5-4" data-line-number="4">                      <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb5-5" data-line-number="5">sim1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/normalize">normalize</a></span>(sim1)</a>
<a class="sourceLine" id="cb5-6" data-line-number="6"><span class="co">#&gt; Warning in .local(object, ...): using library sizes as size factors</span></a>
<a class="sourceLine" id="cb5-7" data-line-number="7"><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plot_reddim">plotPCA</a></span>(sim1, <span class="dt">colour_by =</span> <span class="st">"Batch"</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/labs">ggtitle</a></span>(<span class="st">"Small batch effects"</span>)</a></code></pre></div>
<p><img src="splat_params_files/figure-html/batch-factors-1.png" width="700"></p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"></a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="co"># Simulation with big batch effects</span></a>
<a class="sourceLine" id="cb6-3" data-line-number="3">sim2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/splatSimulate.html">splatSimulate</a></span>(params, <span class="dt">batchCells =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">100</span>, <span class="dv">100</span>),</a>
<a class="sourceLine" id="cb6-4" data-line-number="4">                      <span class="dt">batch.facLoc =</span> <span class="fl">0.5</span>, <span class="dt">batch.facScale =</span> <span class="fl">0.5</span>,</a>
<a class="sourceLine" id="cb6-5" data-line-number="5">                      <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb6-6" data-line-number="6">sim2 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/normalize">normalize</a></span>(sim2)</a>
<a class="sourceLine" id="cb6-7" data-line-number="7"><span class="co">#&gt; Warning in .local(object, ...): using library sizes as size factors</span></a>
<a class="sourceLine" id="cb6-8" data-line-number="8"><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plot_reddim">plotPCA</a></span>(sim2, <span class="dt">colour_by =</span> <span class="st">"Batch"</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/labs">ggtitle</a></span>(<span class="st">"Big batch effects"</span>)</a></code></pre></div>
<p><img src="splat_params_files/figure-html/batch-factors-2.png" width="700"></p>
</div>
</div>
<div id="mean-parameters" class="section level2">
<h2 class="hasAnchor">
<a href="#mean-parameters" class="anchor"></a>Mean parameters</h2>
<p>These parameters control the distribution that is used to generate the underlying original gene means.</p>
<div id="mean-shape---mean-shape-and-mean-rate---mean-rate" class="section level3">
<h3 class="hasAnchor">
<a href="#mean-shape---mean-shape-and-mean-rate---mean-rate" class="anchor"></a><code>mean.shape</code> - Mean shape and <code>mean.rate</code> - Mean rate</h3>
<p>These parameters control the Gamma distribution that gene means are drawn from. The relationship between shape and rate can be complex and it is often better to use values estimated from a real dataset than to try and manually set them. Although these parameters control the base gene means the means in the final simulation will depend upon other parts of the model, particularly the simulated total counts per cell (library sizes).</p>
</div>
</div>
<div id="library-size-parameters" class="section level2">
<h2 class="hasAnchor">
<a href="#library-size-parameters" class="anchor"></a>Library size parameters</h2>
<p>These parameters control the expected number of counts for each cell. Note that because of sampling the actual counts per cell in the final simulation may be different. Turning on the dropout effect will also effect this. We use the term “library size” here for consistency but expected total counts would be more appropriate.</p>
<div id="lib-loc---library-size-location-and-lib-scale---library-size-scale" class="section level3">
<h3 class="hasAnchor">
<a href="#lib-loc---library-size-location-and-lib-scale---library-size-scale" class="anchor"></a><code>lib.loc</code> - Library size location and <code>lib.scale</code> - Library size scale</h3>
<p>These parameters control the shape of the distribution that is used to generate library sizes for each cell. Increasing <code>lib.loc</code> will lead to more counts per cell and increasing <code>lib.scale</code> will result in more variability in the counts per cell.</p>
</div>
<div id="lib-norm---library-size-distribution" class="section level3">
<h3 class="hasAnchor">
<a href="#lib-norm---library-size-distribution" class="anchor"></a><code>lib.norm</code> - Library size distribution</h3>
<p>The default (and recommended) distribution used for library sizes in the Splat simulation is a log-normal. However, in rare cases a normal distribution might be more appropriate. Setting <code>lib.norm</code> to <code>TRUE</code> will use a normal distribution instead of a log-normal.</p>
</div>
</div>
<div id="expression-outlier-parameters" class="section level2">
<h2 class="hasAnchor">
<a href="#expression-outlier-parameters" class="anchor"></a>Expression outlier parameters</h2>
<p>When developing the Splat simulation we found that while the Gamma distribution was generally a good match for gene means for some datasets it did not properly capture highly expressed genes. For this reason we added expression outliers to the Splat model.</p>
<div id="out-prob---expression-outlier-probability" class="section level3">
<h3 class="hasAnchor">
<a href="#out-prob---expression-outlier-probability" class="anchor"></a><code>out.prob</code> - Expression outlier probability</h3>
<p>This parameter controls the probability that genes will be selected to be expression outliers. Higher values will results in more outlier genes.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="co"># Few outliers</span></a>
<a class="sourceLine" id="cb7-2" data-line-number="2">sim1 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/splatSimulate.html">splatSimulate</a></span>(<span class="dt">out.prob =</span> <span class="fl">0.001</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb7-3" data-line-number="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/as.data.frame">as.data.frame</a></span>(<span class="kw">rowData</span>(sim1)),</a>
<a class="sourceLine" id="cb7-4" data-line-number="4">       <span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(<span class="dt">x =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log10</a></span>(GeneMean), <span class="dt">fill =</span> OutlierFactor <span class="op">!=</span><span class="st"> </span><span class="dv">1</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb7-5" data-line-number="5"><span class="st">    </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/geom_histogram">geom_histogram</a></span>(<span class="dt">bins =</span> <span class="dv">100</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb7-6" data-line-number="6"><span class="st">    </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/labs">ggtitle</a></span>(<span class="st">"Few outliers"</span>)</a></code></pre></div>
<p><img src="splat_params_files/figure-html/outlier-prob-1.png" width="700"></p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"></a>
<a class="sourceLine" id="cb8-2" data-line-number="2"><span class="co"># Lots of outliers</span></a>
<a class="sourceLine" id="cb8-3" data-line-number="3">sim2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/splatSimulate.html">splatSimulate</a></span>(<span class="dt">out.prob =</span> <span class="fl">0.2</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb8-4" data-line-number="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/as.data.frame">as.data.frame</a></span>(<span class="kw">rowData</span>(sim2)),</a>
<a class="sourceLine" id="cb8-5" data-line-number="5">       <span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(<span class="dt">x =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log10</a></span>(GeneMean), <span class="dt">fill =</span> OutlierFactor <span class="op">!=</span><span class="st"> </span><span class="dv">1</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb8-6" data-line-number="6"><span class="st">    </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/geom_histogram">geom_histogram</a></span>(<span class="dt">bins =</span> <span class="dv">100</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb8-7" data-line-number="7"><span class="st">    </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/labs">ggtitle</a></span>(<span class="st">"Lots of outliers"</span>)</a></code></pre></div>
<p><img src="splat_params_files/figure-html/outlier-prob-2.png" width="700"></p>
</div>
<div id="out-facloc---expression-outlier-factor-location-and-out-facscale---expression-outlier-factor-scale" class="section level3">
<h3 class="hasAnchor">
<a href="#out-facloc---expression-outlier-factor-location-and-out-facscale---expression-outlier-factor-scale" class="anchor"></a><code>out.facLoc</code> - Expression outlier factor location and <code>out.facScale</code> - Expression outlier factor scale</h3>
<p>The expression outlier factors are drawn from a log-normal distribution controlled by these parameters. The generated factors are applied to the median mean expression rather than the existing mean for the selected genes. This is to be consistent with the estimation procedure for these factors and to make sure that the final means are outliers. For example to avoid the situation where a factor is applied to a lowly expressed gene, making it just moderately expressed rather than an expression outlier.</p>
</div>
<div id="group-parameters" class="section level3">
<h3 class="hasAnchor">
<a href="#group-parameters" class="anchor"></a>Group parameters</h3>
<p>Up until this stage of the simulation only a single population of cells is considered but we often want to simulate datasets with multiple kinds of cells. We do this by assigning cells to groups.</p>
</div>
<div id="ngroups---number-of-groups" class="section level3">
<h3 class="hasAnchor">
<a href="#ngroups---number-of-groups" class="anchor"></a><code>nGroups</code> - Number of groups</h3>
<p>The number of groups to simulate. This parameter cannot be set directly and is controlled using <code>group.prob</code>.</p>
</div>
<div id="group-prob---group-probabilities" class="section level3">
<h3 class="hasAnchor">
<a href="#group-prob---group-probabilities" class="anchor"></a><code>group.prob</code> - Group probabilities</h3>
<p>A vector giving the probability that cells will be assigned to groups. The length of the vector gives the number of groups (<code>nGroups</code>) and the probabilities must sum to 1. Adjusting the number and relative values of the probabilities changes the number and relative sizes of the groups. To simulate groups we also need to use the <code>splatSimulateGroups</code> function or set <code>method = "groups"</code>.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">params.groups &lt;-<span class="st"> </span><span class="kw"><a href="../reference/newParams.html">newSplatParams</a></span>(<span class="dt">batchCells =</span> <span class="dv">500</span>, <span class="dt">nGenes =</span> <span class="dv">1000</span>)</a>
<a class="sourceLine" id="cb9-2" data-line-number="2"></a>
<a class="sourceLine" id="cb9-3" data-line-number="3"><span class="co"># One small group, one big group</span></a>
<a class="sourceLine" id="cb9-4" data-line-number="4">sim1 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/splatSimulate.html">splatSimulateGroups</a></span>(params.groups, <span class="dt">group.prob =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.9</span>, <span class="fl">0.1</span>),</a>
<a class="sourceLine" id="cb9-5" data-line-number="5">                            <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb9-6" data-line-number="6">sim1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/normalize">normalize</a></span>(sim1)</a>
<a class="sourceLine" id="cb9-7" data-line-number="7"><span class="co">#&gt; Warning in .local(object, ...): using library sizes as size factors</span></a>
<a class="sourceLine" id="cb9-8" data-line-number="8"><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plot_reddim">plotPCA</a></span>(sim1, <span class="dt">colour_by =</span> <span class="st">"Group"</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/labs">ggtitle</a></span>(<span class="st">"One small group, one big group"</span>)</a></code></pre></div>
<p><img src="splat_params_files/figure-html/groups-1.png" width="700"></p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"></a>
<a class="sourceLine" id="cb10-2" data-line-number="2"><span class="co"># Five groups</span></a>
<a class="sourceLine" id="cb10-3" data-line-number="3">sim2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/splatSimulate.html">splatSimulateGroups</a></span>(params.groups,</a>
<a class="sourceLine" id="cb10-4" data-line-number="4">                            <span class="dt">group.prob =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>),</a>
<a class="sourceLine" id="cb10-5" data-line-number="5">                            <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb10-6" data-line-number="6">sim2 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/normalize">normalize</a></span>(sim2)</a>
<a class="sourceLine" id="cb10-7" data-line-number="7"><span class="co">#&gt; Warning in .local(object, ...): using library sizes as size factors</span></a>
<a class="sourceLine" id="cb10-8" data-line-number="8"><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plot_reddim">plotPCA</a></span>(sim2, <span class="dt">colour_by =</span> <span class="st">"Group"</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/labs">ggtitle</a></span>(<span class="st">"Five groups"</span>)</a></code></pre></div>
<p><img src="splat_params_files/figure-html/groups-2.png" width="700"></p>
<p><strong>Note:</strong> Once there are more than three or four groups it becomes difficult to properly view them in PCA space. We use PCA here for simplicity but generally a non-linear dimensionality reduction such as t-SNE or UMAP is a more useful way to visualise the groups.</p>
</div>
<div id="differential-expression-parameters" class="section level3">
<h3 class="hasAnchor">
<a href="#differential-expression-parameters" class="anchor"></a>Differential expression parameters</h3>
<p>Different groups are created by modifying the base expression levels of selected genes. The process for doing this is to simulate differential expression (DE) between each group and a fictional base cell. Altering the differential expression parameters controls how similar groups are to each other.</p>
</div>
<div id="de-prob---de-probability" class="section level3">
<h3 class="hasAnchor">
<a href="#de-prob---de-probability" class="anchor"></a><code>de.prob</code> - DE probability</h3>
<p>This parameter controls the probability that a gene will be selected to be differentially expressed.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="co"># Few DE genes</span></a>
<a class="sourceLine" id="cb11-2" data-line-number="2">sim1 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/splatSimulate.html">splatSimulateGroups</a></span>(params.groups, <span class="dt">group.prob =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.5</span>, <span class="fl">0.5</span>),</a>
<a class="sourceLine" id="cb11-3" data-line-number="3">                            <span class="dt">de.prob =</span> <span class="fl">0.01</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb11-4" data-line-number="4">sim1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/normalize">normalize</a></span>(sim1)</a>
<a class="sourceLine" id="cb11-5" data-line-number="5"><span class="co">#&gt; Warning in .local(object, ...): using library sizes as size factors</span></a>
<a class="sourceLine" id="cb11-6" data-line-number="6"><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plot_reddim">plotPCA</a></span>(sim1, <span class="dt">colour_by =</span> <span class="st">"Group"</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/labs">ggtitle</a></span>(<span class="st">"Few DE genes"</span>)</a></code></pre></div>
<p><img src="splat_params_files/figure-html/de-prob-1.png" width="700"></p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"></a>
<a class="sourceLine" id="cb12-2" data-line-number="2"><span class="co"># Lots of DE genes</span></a>
<a class="sourceLine" id="cb12-3" data-line-number="3">sim2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/splatSimulate.html">splatSimulateGroups</a></span>(params.groups, <span class="dt">group.prob =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.5</span>, <span class="fl">0.5</span>),</a>
<a class="sourceLine" id="cb12-4" data-line-number="4">                            <span class="dt">de.prob =</span> <span class="fl">0.3</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb12-5" data-line-number="5">sim2 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/normalize">normalize</a></span>(sim2)</a>
<a class="sourceLine" id="cb12-6" data-line-number="6"><span class="co">#&gt; Warning in .local(object, ...): using library sizes as size factors</span></a>
<a class="sourceLine" id="cb12-7" data-line-number="7"><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plot_reddim">plotPCA</a></span>(sim2, <span class="dt">colour_by =</span> <span class="st">"Group"</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/labs">ggtitle</a></span>(<span class="st">"Lots of DE genes"</span>)</a></code></pre></div>
<p><img src="splat_params_files/figure-html/de-prob-2.png" width="700"></p>
</div>
<div id="de-downprob---down-regulation-probability" class="section level3">
<h3 class="hasAnchor">
<a href="#de-downprob---down-regulation-probability" class="anchor"></a><code>de.downProb</code> - Down-regulation probability</h3>
<p>A selected DE gene can be either down-regulated (has a factor less than one) or up-regulated (has a factor greater than one). This parameter controls the probability that a selected gene will be down-regulated.</p>
</div>
<div id="de-facloc---de-factor-location-and-de-facscale---de-factor-scale" class="section level3">
<h3 class="hasAnchor">
<a href="#de-facloc---de-factor-location-and-de-facscale---de-factor-scale" class="anchor"></a><code>de.facLoc</code> - DE factor location and <code>de.facScale</code> - DE factor scale</h3>
<p>Differential expression factors are produced from a log-normal distribution in a similar way to batch effect factors and expression outlier factors. Changing these parameters can result in more or less extreme differences between groups.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1"><span class="co"># Small DE factors</span></a>
<a class="sourceLine" id="cb13-2" data-line-number="2">sim1 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/splatSimulate.html">splatSimulateGroups</a></span>(params.groups, <span class="dt">group.prob =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.5</span>, <span class="fl">0.5</span>),</a>
<a class="sourceLine" id="cb13-3" data-line-number="3">                            <span class="dt">de.facLoc =</span> <span class="fl">0.01</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb13-4" data-line-number="4">sim1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/normalize">normalize</a></span>(sim1)</a>
<a class="sourceLine" id="cb13-5" data-line-number="5"><span class="co">#&gt; Warning in .local(object, ...): using library sizes as size factors</span></a>
<a class="sourceLine" id="cb13-6" data-line-number="6"><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plot_reddim">plotPCA</a></span>(sim1, <span class="dt">colour_by =</span> <span class="st">"Group"</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/labs">ggtitle</a></span>(<span class="st">"Small DE factors"</span>)</a></code></pre></div>
<p><img src="splat_params_files/figure-html/de-factors-1.png" width="700"></p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"></a>
<a class="sourceLine" id="cb14-2" data-line-number="2"><span class="co"># Big DE factors</span></a>
<a class="sourceLine" id="cb14-3" data-line-number="3">sim2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/splatSimulate.html">splatSimulateGroups</a></span>(params.groups, <span class="dt">group.prob =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.5</span>, <span class="fl">0.5</span>),</a>
<a class="sourceLine" id="cb14-4" data-line-number="4">                            <span class="dt">de.facLoc =</span> <span class="fl">0.3</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb14-5" data-line-number="5">sim2 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/normalize">normalize</a></span>(sim2)</a>
<a class="sourceLine" id="cb14-6" data-line-number="6"><span class="co">#&gt; Warning in .local(object, ...): using library sizes as size factors</span></a>
<a class="sourceLine" id="cb14-7" data-line-number="7"><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plot_reddim">plotPCA</a></span>(sim2, <span class="dt">colour_by =</span> <span class="st">"Group"</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/labs">ggtitle</a></span>(<span class="st">"Big DE factors"</span>)</a></code></pre></div>
<p><img src="splat_params_files/figure-html/de-factors-2.png" width="700"></p>
<p>Just looking at the PCA plots this effect seems similar to adjusting <code>de.prob</code> but the effect is achieved in a different way. A higher <code>de.prob</code> means that more genes are differentially expressed but changing the DE factors changes the level of DE for the same number of genes.</p>
</div>
<div id="complex-differential-expression" class="section level3">
<h3 class="hasAnchor">
<a href="#complex-differential-expression" class="anchor"></a>Complex differential expression</h3>
<p>Each of the differential expression parameters can be specified for each group by providing a vector of values. These vectors must be the same length as <code>group.prob</code>. Specifying parameters as vectors allows more complex simulations where groups are more or less different to each other rather than being equally distinct. Here are some examples of different DE scenarios.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1"><span class="co"># Different DE probs</span></a>
<a class="sourceLine" id="cb15-2" data-line-number="2">sim1 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/splatSimulate.html">splatSimulateGroups</a></span>(params.groups,</a>
<a class="sourceLine" id="cb15-3" data-line-number="3">                            <span class="dt">group.prob =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>),</a>
<a class="sourceLine" id="cb15-4" data-line-number="4">                            <span class="dt">de.prob =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.01</span>, <span class="fl">0.01</span>, <span class="fl">0.1</span>, <span class="fl">0.1</span>, <span class="fl">0.3</span>),</a>
<a class="sourceLine" id="cb15-5" data-line-number="5">                            <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb15-6" data-line-number="6">sim1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/normalize">normalize</a></span>(sim1)</a>
<a class="sourceLine" id="cb15-7" data-line-number="7"><span class="co">#&gt; Warning in .local(object, ...): using library sizes as size factors</span></a>
<a class="sourceLine" id="cb15-8" data-line-number="8"><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plot_reddim">plotPCA</a></span>(sim1, <span class="dt">colour_by =</span> <span class="st">"Group"</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb15-9" data-line-number="9"><span class="st">    </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/labs">labs</a></span>(<span class="dt">title =</span> <span class="st">"Different DE probabilities"</span>,</a>
<a class="sourceLine" id="cb15-10" data-line-number="10">         <span class="dt">caption =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="st">"Groups 1 and 2 have very few DE genes,"</span>,</a>
<a class="sourceLine" id="cb15-11" data-line-number="11">                         <span class="st">"Groups 3 and 4 have the default number,"</span>,</a>
<a class="sourceLine" id="cb15-12" data-line-number="12">                         <span class="st">"Group 5 has many DE genes"</span>))</a></code></pre></div>
<p><img src="splat_params_files/figure-html/complex-de-1.png" width="700"></p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1">    </a>
<a class="sourceLine" id="cb16-2" data-line-number="2"></a>
<a class="sourceLine" id="cb16-3" data-line-number="3"><span class="co"># Different DE factors</span></a>
<a class="sourceLine" id="cb16-4" data-line-number="4">sim2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/splatSimulate.html">splatSimulateGroups</a></span>(params.groups,</a>
<a class="sourceLine" id="cb16-5" data-line-number="5">                            <span class="dt">group.prob =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>),</a>
<a class="sourceLine" id="cb16-6" data-line-number="6">                            <span class="dt">de.facLoc =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.01</span>, <span class="fl">0.01</span>, <span class="fl">0.1</span>, <span class="fl">0.1</span>, <span class="fl">0.2</span>),</a>
<a class="sourceLine" id="cb16-7" data-line-number="7">                            <span class="dt">de.facScale =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.2</span>, <span class="fl">0.5</span>, <span class="fl">0.2</span>, <span class="fl">0.5</span>, <span class="fl">0.4</span>),</a>
<a class="sourceLine" id="cb16-8" data-line-number="8">                            <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb16-9" data-line-number="9">sim2 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/normalize">normalize</a></span>(sim2)</a>
<a class="sourceLine" id="cb16-10" data-line-number="10"><span class="co">#&gt; Warning in .local(object, ...): using library sizes as size factors</span></a>
<a class="sourceLine" id="cb16-11" data-line-number="11"><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plot_reddim">plotPCA</a></span>(sim2, <span class="dt">colour_by =</span> <span class="st">"Group"</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb16-12" data-line-number="12"><span class="st">    </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/labs">labs</a></span>(<span class="dt">title =</span> <span class="st">"Different DE factors"</span>,</a>
<a class="sourceLine" id="cb16-13" data-line-number="13">         <span class="dt">caption =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="st">"Group 1 has factors with small location (value),"</span>,</a>
<a class="sourceLine" id="cb16-14" data-line-number="14">                         <span class="st">"and scale (variability),"</span>,</a>
<a class="sourceLine" id="cb16-15" data-line-number="15">                         <span class="st">"Group 2 has small location and greater scale.</span><span class="ch">\n</span><span class="st">"</span>,</a>
<a class="sourceLine" id="cb16-16" data-line-number="16">                         <span class="st">"Groups 3 and 4 have greater location with small,"</span>,</a>
<a class="sourceLine" id="cb16-17" data-line-number="17">                         <span class="st">"and large scales"</span>,</a>
<a class="sourceLine" id="cb16-18" data-line-number="18">                         <span class="st">"Group 5 has bigger factors with moderate variability"</span>))</a></code></pre></div>
<p><img src="splat_params_files/figure-html/complex-de-2.png" width="700"></p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"></a>
<a class="sourceLine" id="cb17-2" data-line-number="2"><span class="co"># Combination of everything</span></a>
<a class="sourceLine" id="cb17-3" data-line-number="3">sim3 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/splatSimulate.html">splatSimulateGroups</a></span>(params.groups,</a>
<a class="sourceLine" id="cb17-4" data-line-number="4">                            <span class="dt">group.prob =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.05</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.35</span>),</a>
<a class="sourceLine" id="cb17-5" data-line-number="5">                            <span class="dt">de.prob =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.3</span>, <span class="fl">0.1</span>, <span class="fl">0.2</span>, <span class="fl">0.01</span>, <span class="fl">0.1</span>),</a>
<a class="sourceLine" id="cb17-6" data-line-number="6">                            <span class="dt">de.downProb =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.1</span>, <span class="fl">0.4</span>, <span class="fl">0.9</span>, <span class="fl">0.6</span>, <span class="fl">0.5</span>),</a>
<a class="sourceLine" id="cb17-7" data-line-number="7">                            <span class="dt">de.facLoc =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.6</span>, <span class="fl">0.1</span>, <span class="fl">0.1</span>, <span class="fl">0.01</span>, <span class="fl">0.2</span>),</a>
<a class="sourceLine" id="cb17-8" data-line-number="8">                            <span class="dt">de.facScale =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.1</span>, <span class="fl">0.4</span>, <span class="fl">0.2</span>, <span class="fl">0.5</span>, <span class="fl">0.4</span>),</a>
<a class="sourceLine" id="cb17-9" data-line-number="9">                            <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb17-10" data-line-number="10">sim3 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/normalize">normalize</a></span>(sim3)</a>
<a class="sourceLine" id="cb17-11" data-line-number="11"><span class="co">#&gt; Warning in .local(object, ...): using library sizes as size factors</span></a>
<a class="sourceLine" id="cb17-12" data-line-number="12"><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plot_reddim">plotPCA</a></span>(sim3, <span class="dt">colour_by =</span> <span class="st">"Group"</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb17-13" data-line-number="13"><span class="st">    </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/labs">labs</a></span>(<span class="dt">title =</span> <span class="st">"Different DE factors"</span>,</a>
<a class="sourceLine" id="cb17-14" data-line-number="14">         <span class="dt">caption =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(</a>
<a class="sourceLine" id="cb17-15" data-line-number="15">             <span class="st">"Group 1 is small with many very up-regulated DE genes,"</span>,</a>
<a class="sourceLine" id="cb17-16" data-line-number="16">             <span class="st">"Group 2 has the default DE parameters,</span><span class="ch">\n</span><span class="st">"</span>,</a>
<a class="sourceLine" id="cb17-17" data-line-number="17">             <span class="st">"Group 3 has many down-regulated DE genes,"</span>,</a>
<a class="sourceLine" id="cb17-18" data-line-number="18">             <span class="st">"Group 4 has very few DE genes,"</span>,</a>
<a class="sourceLine" id="cb17-19" data-line-number="19">             <span class="st">"Group 5 is large with moderate DE factors"</span>)</a>
<a class="sourceLine" id="cb17-20" data-line-number="20">         )</a></code></pre></div>
<p><img src="splat_params_files/figure-html/complex-de-3.png" width="700"></p>
</div>
</div>
<div id="biological-coefficient-of-variation-bcv-parameters" class="section level2">
<h2 class="hasAnchor">
<a href="#biological-coefficient-of-variation-bcv-parameters" class="anchor"></a>Biological Coefficient of Variation (BCV) parameters</h2>
<p>The BCV parameters control the variability of the genes in the simulated dataset.</p>
<div id="bcv-common---common-bcv" class="section level3">
<h3 class="hasAnchor">
<a href="#bcv-common---common-bcv" class="anchor"></a><code>bcv.common</code> - Common BCV</h3>
<p>The <code>bcv.common</code> parameter controls the underlying common variability across all genes in the dataset.</p>
</div>
<div id="bcv-df---bcv-degrees-of-freedom" class="section level3">
<h3 class="hasAnchor">
<a href="#bcv-df---bcv-degrees-of-freedom" class="anchor"></a><code>bcv.df</code> - BCV Degrees of Freedom</h3>
<p>This parameter sets the degrees of freedom used in the BCV inverse chi-squared distribution. Changing this changes the effect of mean expression on the variability of a gene.</p>
</div>
</div>
<div id="dropout-parameters" class="section level2">
<h2 class="hasAnchor">
<a href="#dropout-parameters" class="anchor"></a>Dropout parameters</h2>
<p>These parameters control whether additional dropout is added to increase the number of zeros in the simulated dataset and if it is how that is applied.</p>
<div id="dropout-type---dropout-type" class="section level3">
<h3 class="hasAnchor">
<a href="#dropout-type---dropout-type" class="anchor"></a><code>dropout.type</code> - Dropout type</h3>
<p>This parameter determines the kind of dropout effect to simulate. Setting it to <code>"none"</code> means no dropout, “<code>experiment</code>” is global dropout using the same set of parameters for every cell, <code>"batch"</code> uses the same parameters for every cell in the same batch, <code>"group"</code> uses the same parameters for every cell in the same group and <code>"cell"</code> uses a different set of parameters for every cell.</p>
</div>
<div id="dropout-mid---dropout-mid-point-and-dropout-shape---dropout-shape" class="section level3">
<h3 class="hasAnchor">
<a href="#dropout-mid---dropout-mid-point-and-dropout-shape---dropout-shape" class="anchor"></a><code>dropout.mid</code> - Dropout mid point and <code>dropout.shape</code> - Dropout shape</h3>
<p>The probability that a particular count in a particular cell is set to zero is related to the mean expression of that gene in that cell. This relationship is represented using a logistic function with these parameters. The <code>dropout.mid</code> parameter control the point at which the probability is equal to 0.5 and the <code>dropout.shape</code> controls how it changes with increasing expression. These parameters must be vectors of the appropriate length depending on the selected dropout type.</p>
</div>
</div>
<div id="path-parameters" class="section level2">
<h2 class="hasAnchor">
<a href="#path-parameters" class="anchor"></a>Path parameters</h2>
<p>For many uses simulating groups is sufficient but in some cases it is more appropriate to simulate continuous changes between cell types. The number of paths and the probability of assigning cells to them is still controlled by the <code>group.prob</code> parameter and the amount of change along a path is controlled by the DE parameters but other aspects of the <code>splatSimulatePaths</code> model are controlled by these parameters.</p>
<div id="path-from---path-origin" class="section level3">
<h3 class="hasAnchor">
<a href="#path-from---path-origin" class="anchor"></a><code>path.from</code> - Path origin</h3>
<p>This parameter controls the order of differentiation paths. It is a vector the same length as <code>group.prob</code> giving the starting position of each path. For example a <code>path.from</code> of <code><a href="https://www.rdocumentation.org/packages/base/topics/c">c(0, 1, 1, 3)</a></code> would indicate the Path 1 starts at the origin (0), Path 2 starts at the end of Path 1, Path 3 also starts at the end of Path 1 (a branch point) and Path 4 starts at the end of Path 3.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1"><span class="co"># Linear paths</span></a>
<a class="sourceLine" id="cb18-2" data-line-number="2">sim1 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/splatSimulate.html">splatSimulatePaths</a></span>(params.groups,</a>
<a class="sourceLine" id="cb18-3" data-line-number="3">                           <span class="dt">group.prob =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.25</span>, <span class="fl">0.25</span>, <span class="fl">0.25</span>, <span class="fl">0.25</span>),</a>
<a class="sourceLine" id="cb18-4" data-line-number="4">                           <span class="dt">de.prob =</span> <span class="fl">0.5</span>, <span class="dt">de.facLoc =</span> <span class="fl">0.2</span>,</a>
<a class="sourceLine" id="cb18-5" data-line-number="5">                           <span class="dt">path.from =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>),</a>
<a class="sourceLine" id="cb18-6" data-line-number="6">                           <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb18-7" data-line-number="7">sim1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/normalize">normalize</a></span>(sim1)</a>
<a class="sourceLine" id="cb18-8" data-line-number="8"><span class="co">#&gt; Warning in .local(object, ...): using library sizes as size factors</span></a>
<a class="sourceLine" id="cb18-9" data-line-number="9"><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plot_reddim">plotPCA</a></span>(sim1, <span class="dt">colour_by =</span> <span class="st">"Group"</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/labs">ggtitle</a></span>(<span class="st">"Linear paths"</span>)</a></code></pre></div>
<p><img src="splat_params_files/figure-html/paths-1.png" width="700"></p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"></a>
<a class="sourceLine" id="cb19-2" data-line-number="2"><span class="co"># Branching path</span></a>
<a class="sourceLine" id="cb19-3" data-line-number="3">sim2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/splatSimulate.html">splatSimulatePaths</a></span>(params.groups,</a>
<a class="sourceLine" id="cb19-4" data-line-number="4">                           <span class="dt">group.prob =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.25</span>, <span class="fl">0.25</span>, <span class="fl">0.25</span>, <span class="fl">0.25</span>),</a>
<a class="sourceLine" id="cb19-5" data-line-number="5">                           <span class="dt">de.prob =</span> <span class="fl">0.5</span>, <span class="dt">de.facLoc =</span> <span class="fl">0.2</span>,</a>
<a class="sourceLine" id="cb19-6" data-line-number="6">                           <span class="dt">path.from =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">3</span>),</a>
<a class="sourceLine" id="cb19-7" data-line-number="7">                           <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb19-8" data-line-number="8">sim2 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/normalize">normalize</a></span>(sim2)</a>
<a class="sourceLine" id="cb19-9" data-line-number="9"><span class="co">#&gt; Warning in .local(object, ...): using library sizes as size factors</span></a>
<a class="sourceLine" id="cb19-10" data-line-number="10"><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plot_reddim">plotPCA</a></span>(sim2, <span class="dt">colour_by =</span> <span class="st">"Group"</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/labs">ggtitle</a></span>(<span class="st">"Branching path"</span>)</a></code></pre></div>
<p><img src="splat_params_files/figure-html/paths-2.png" width="700"></p>
</div>
<div id="path-nsteps---number-of-steps" class="section level3">
<h3 class="hasAnchor">
<a href="#path-nsteps---number-of-steps" class="anchor"></a><code>path.nSteps</code> - Number of steps</h3>
<p>A path is created by using the same differential expression procedure as used for groups to generate an end point. Interpolation is then used to create a series of steps between the start and end points. This parameter controls the number of steps along a path and therefore how discrete or smooth it is.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1"><span class="co"># Few steps</span></a>
<a class="sourceLine" id="cb20-2" data-line-number="2">sim1 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/splatSimulate.html">splatSimulatePaths</a></span>(params.groups, <span class="dt">path.nSteps =</span> <span class="dv">3</span>,</a>
<a class="sourceLine" id="cb20-3" data-line-number="3">                           <span class="dt">de.prob =</span> <span class="fl">0.5</span>, <span class="dt">de.facLoc =</span> <span class="fl">0.2</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb20-4" data-line-number="4">sim1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/normalize">normalize</a></span>(sim1)</a>
<a class="sourceLine" id="cb20-5" data-line-number="5"><span class="co">#&gt; Warning in .local(object, ...): using library sizes as size factors</span></a>
<a class="sourceLine" id="cb20-6" data-line-number="6"><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plot_reddim">plotPCA</a></span>(sim1, <span class="dt">colour_by =</span> <span class="st">"Step"</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/labs">ggtitle</a></span>(<span class="st">"Few steps"</span>)</a></code></pre></div>
<p><img src="splat_params_files/figure-html/paths-steps-1.png" width="700"></p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1"></a>
<a class="sourceLine" id="cb21-2" data-line-number="2"><span class="co"># Lots of steps</span></a>
<a class="sourceLine" id="cb21-3" data-line-number="3">sim2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/splatSimulate.html">splatSimulatePaths</a></span>(params.groups, <span class="dt">path.nSteps =</span> <span class="dv">1000</span>,</a>
<a class="sourceLine" id="cb21-4" data-line-number="4">                           <span class="dt">de.prob =</span> <span class="fl">0.5</span>, <span class="dt">de.facLoc =</span> <span class="fl">0.2</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb21-5" data-line-number="5">sim2 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/normalize">normalize</a></span>(sim2)</a>
<a class="sourceLine" id="cb21-6" data-line-number="6"><span class="co">#&gt; Warning in .local(object, ...): using library sizes as size factors</span></a>
<a class="sourceLine" id="cb21-7" data-line-number="7"><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plot_reddim">plotPCA</a></span>(sim2, <span class="dt">colour_by =</span> <span class="st">"Step"</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/labs">ggtitle</a></span>(<span class="st">"Lots of steps"</span>)</a></code></pre></div>
<p><img src="splat_params_files/figure-html/paths-steps-2.png" width="700"></p>
</div>
<div id="path-skew---path-skew" class="section level3">
<h3 class="hasAnchor">
<a href="#path-skew---path-skew" class="anchor"></a><code>path.skew</code> - Path skew</h3>
<p>By default cells are evenly distributed along a path but it sometimes be useful to introduce a skew in the distribution, For example you may want to simulate a scenario with few stem-like cells and many differentiated cells. Setting <code>path.skew</code> to 0 will mean that all cells come from the end point while higher values up to 1 will skew them towards the start point.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1"><span class="co"># Skew towards the end</span></a>
<a class="sourceLine" id="cb22-2" data-line-number="2">sim1 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/splatSimulate.html">splatSimulatePaths</a></span>(params.groups, <span class="dt">path.skew =</span> <span class="fl">0.1</span>,</a>
<a class="sourceLine" id="cb22-3" data-line-number="3">                           <span class="dt">de.prob =</span> <span class="fl">0.5</span>, <span class="dt">de.facLoc =</span> <span class="fl">0.2</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb22-4" data-line-number="4">sim1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/normalize">normalize</a></span>(sim1)</a>
<a class="sourceLine" id="cb22-5" data-line-number="5"><span class="co">#&gt; Warning in .local(object, ...): using library sizes as size factors</span></a>
<a class="sourceLine" id="cb22-6" data-line-number="6"><span class="kw"><a href="https://www.rdocumentation.org/packages/scater/topics/plot_reddim">plotPCA</a></span>(sim1, <span class="dt">colour_by =</span> <span class="st">"Step"</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/labs">ggtitle</a></span>(<span class="st">"Skewed towards the end"</span>)</a></code></pre></div>
<p><img src="splat_params_files/figure-html/paths-skew-1.png" width="700"></p>
</div>
<div id="path-nonlinearprob---non-linear-probability" class="section level3">
<h3 class="hasAnchor">
<a href="#path-nonlinearprob---non-linear-probability" class="anchor"></a><code>path.nonlinearProb</code> - Non-linear probability</h3>
<p>Most genes are interpolated in a linear way along a path but in reality this may not always be the case. For example it is easy to imagine a gene that is lowly-expressed at the start of a process, highly-expressed in the middle and lowly-expressed again at the end. The <code>path.nonlinearProb</code> parameter controls the probability that a gene will change in a non-linear way along a path.</p>
</div>
<div id="path-sigmafac---path-skew" class="section level3">
<h3 class="hasAnchor">
<a href="#path-sigmafac---path-skew" class="anchor"></a><code>path.sigmaFac</code> - Path skew</h3>
<p>Non-linear changes along a path are achieved by building a Brownian bridge between the two end points. A Brownian bridge is Brownian motion controlled in such a way that the end points are fixed. The <code>path.sigmaFac</code> parameter controls how extreme each step in the Brownian motion is and therefore how much the interpolation differs from a linear path.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#the-base-splat-model">The base Splat model</a></li>
      <li>
<a href="#splat-simulation-parameters">Splat simulation parameters</a><ul class="nav nav-pills nav-stacked">
<li><a href="#global-parameters">Global parameters</a></li>
      <li><a href="#batch-parameters">Batch parameters</a></li>
      <li><a href="#mean-parameters">Mean parameters</a></li>
      <li><a href="#library-size-parameters">Library size parameters</a></li>
      <li><a href="#expression-outlier-parameters">Expression outlier parameters</a></li>
      <li><a href="#biological-coefficient-of-variation-bcv-parameters">Biological Coefficient of Variation (BCV) parameters</a></li>
      <li><a href="#dropout-parameters">Dropout parameters</a></li>
      <li><a href="#path-parameters">Path parameters</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Luke Zappia, Belinda Phipson, Alicia Oshlack.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
