<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Splat simulation — splatSimulate • Splatter</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Splat simulation — splatSimulate" />

<meta property="og:description" content="Simulate count data from a fictional single-cell RNA-seq experiment using
the Splat method." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-52309538-4"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-52309538-4');
</script>

  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Splatter</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">1.5.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/splatter.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/Oshlack/splatter">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Splat simulation</h1>
    <small class="dont-index">Source: <a href='https://github.com/Oshlack/splatter/blob/master/R/splat-simulate.R'><code>R/splat-simulate.R</code></a></small>
    <div class="hidden name"><code>splatSimulate.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Simulate count data from a fictional single-cell RNA-seq experiment using
the Splat method.</p>
    
    </div>

    <pre class="usage"><span class='fu'>splatSimulate</span>(<span class='kw'>params</span> <span class='kw'>=</span> <span class='fu'><a href='newParams.html'>newSplatParams</a></span>(), <span class='kw'>method</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"single"</span>, <span class='st'>"groups"</span>,
  <span class='st'>"paths"</span>), <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='no'>...</span>)

<span class='fu'>splatSimulateSingle</span>(<span class='kw'>params</span> <span class='kw'>=</span> <span class='fu'><a href='newParams.html'>newSplatParams</a></span>(), <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='no'>...</span>)

<span class='fu'>splatSimulateGroups</span>(<span class='kw'>params</span> <span class='kw'>=</span> <span class='fu'><a href='newParams.html'>newSplatParams</a></span>(), <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='no'>...</span>)

<span class='fu'>splatSimulatePaths</span>(<span class='kw'>params</span> <span class='kw'>=</span> <span class='fu'><a href='newParams.html'>newSplatParams</a></span>(), <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>params</th>
      <td><p>SplatParams object containing parameters for the simulation.
See <code><a href='SplatParams.html'>SplatParams</a></code> for details.</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>which simulation method to use. Options are "single" which
produces a single population, "groups" which produces distinct groups
(eg. cell types) or "paths" which selects cells from continuous
trajectories (eg. differentiation processes).</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>logical. Whether to print progress messages.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>any additional parameter settings to override what is provided in
<code>params</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>SingleCellExperiment object containing the simulated counts and
intermediate values.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Parameters can be set in a variety of ways. If no parameters are provided
the default parameters are used. Any parameters in <code>params</code> can be
overridden by supplying additional arguments through a call to
<code><a href='setParams.html'>setParams</a></code>. This design allows the user flexibility in
how they supply parameters and allows small adjustments without creating a
new <code>SplatParams</code> object. See examples for a demonstration of how this
can be used.</p>
<p>The simulation involves the following steps:</p><ol>
<li><p>Set up simulation object</p></li>
<li><p>Simulate library sizes</p></li>
<li><p>Simulate gene means</p></li>
<li><p>Simulate groups/paths</p></li>
<li><p>Simulate BCV adjusted cell means</p></li>
<li><p>Simulate true counts</p></li>
<li><p>Simulate dropout</p></li>
<li><p>Create final dataset</p></li>
</ol>
    <p>The final output is a
<code><a href='http://www.rdocumentation.org/packages/SingleCellExperiment/topics/SingleCellExperiment'>SingleCellExperiment</a></code> object that
contains the simulated counts but also the values for various intermediate
steps. These are stored in the <code>colData</code> (for cell specific
information), <code>rowData</code> (for gene specific information) or
<code>assays</code> (for gene by cell matrices) slots. This additional
information includes:</p><dl class='dl-horizontal'>
    <dt><code>phenoData</code></dt><dd><p><dl class='dl-horizontal'>
            <dt>Cell</dt><dd><p>Unique cell identifier.</p></dd>
            <dt>Group</dt><dd><p>The group or path the cell belongs to.</p></dd>
            <dt>ExpLibSize</dt><dd><p>The expected library size for that cell.</p></dd>
            <dt>Step (paths only)</dt><dd><p>how far along the path each cell is.</p></dd>
        </dl></p></dd>
    <dt><code>featureData</code></dt><dd><p><dl class='dl-horizontal'>
            <dt>Gene</dt><dd><p>Unique gene identifier.</p></dd>
            <dt>BaseGeneMean</dt><dd><p>The base expression level for that gene.</p></dd>
            <dt>OutlierFactor</dt><dd><p>Expression outlier factor for that gene.
            Values of 1 indicate the gene is not an expression outlier.</p></dd>
            <dt>GeneMean</dt><dd><p>Expression level after applying outlier factors.</p></dd>
            <dt>BatchFac[Batch]</dt><dd><p>The batch effects factor for each gene for
            a particular batch.</p></dd>
            <dt>DEFac[Group]</dt><dd><p>The differential expression factor for each
            gene in a particular group. Values of 1 indicate the gene is not
            differentially expressed.</p></dd>
            <dt>SigmaFac[Path]</dt><dd><p>Factor applied to genes that have
            non-linear changes in expression along a path.</p></dd>
        </dl></p></dd>
    <dt><code>assayData</code></dt><dd><p><dl class='dl-horizontal'>
            <dt>BatchCellMeans</dt><dd><p>The mean expression of genes in each cell
            after adding batch effects.</p></dd>
            <dt>BaseCellMeans</dt><dd><p>The mean expression of genes in each cell
            after any differential expression and adjusted for expected
            library size.</p></dd>
            <dt>BCV</dt><dd><p>The Biological Coefficient of Variation for each gene
            in each cell.</p></dd>
            <dt>CellMeans</dt><dd><p>The mean expression level of genes in each cell
            adjusted for BCV.</p></dd>
            <dt>TrueCounts</dt><dd><p>The simulated counts before dropout.</p></dd>
            <dt>Dropout</dt><dd><p>Logical matrix showing which values have been
            dropped in which cells.</p></dd>
        </dl></p></dd>
</dl>
    <p>Values that have been added by Splatter are named using <code>UpperCamelCase</code>
in order to differentiate them from the values added by analysis packages
which typically use <code>underscore_naming</code>.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Zappia L, Phipson B, Oshlack A. Splatter: simulation of single-cell RNA
sequencing data. Genome Biology (2017).</p>
<p>Paper: <a href='10.1186/s13059-017-1305-0'>10.1186/s13059-017-1305-0</a></p>
<p>Code: <a href='https://github.com/Oshlack/splatter'>https://github.com/Oshlack/splatter</a></p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='splatSimLibSizes.html'>splatSimLibSizes</a></code>, <code><a href='splatSimGeneMeans.html'>splatSimGeneMeans</a></code>,
<code><a href='splatSimBatchEffects.html'>splatSimBatchEffects</a></code>, <code><a href='splatSimBatchCellMeans.html'>splatSimBatchCellMeans</a></code>,
<code><a href='splatSimDE.html'>splatSimDE</a></code>, <code><a href='splatSimCellMeans.html'>splatSimCellMeans</a></code>,
<code><a href='splatSimBCVMeans.html'>splatSimBCVMeans</a></code>, <code><a href='splatSimTrueCounts.html'>splatSimTrueCounts</a></code>,
<code><a href='splatSimDropout.html'>splatSimDropout</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Simulation with default parameters</span>
<span class='no'>sim</span> <span class='kw'>&lt;-</span> <span class='fu'>splatSimulate</span>()</div><div class='output co'>#&gt; <span class='message'>Getting parameters...</span></div><div class='output co'>#&gt; <span class='message'>Creating simulation object...</span></div><div class='output co'>#&gt; <span class='message'>Simulating library sizes...</span></div><div class='output co'>#&gt; <span class='message'>Simulating gene means...</span></div><div class='output co'>#&gt; <span class='message'>Simulating BCV...</span></div><div class='output co'>#&gt; <span class='message'>Simulating counts...</span></div><div class='output co'>#&gt; <span class='message'>Simulating dropout (if needed)...</span></div><div class='output co'>#&gt; <span class='message'>Done!</span></div><span class='co'># NOT RUN {</span>
<span class='co'># Simulation with different number of genes</span>
<span class='no'>sim</span> <span class='kw'>&lt;-</span> <span class='fu'>splatSimulate</span>(<span class='kw'>nGenes</span> <span class='kw'>=</span> <span class='fl'>1000</span>)
<span class='co'># Simulation with custom parameters</span>
<span class='no'>params</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='newParams.html'>newSplatParams</a></span>(<span class='kw'>nGenes</span> <span class='kw'>=</span> <span class='fl'>100</span>, <span class='kw'>mean.rate</span> <span class='kw'>=</span> <span class='fl'>0.5</span>)
<span class='no'>sim</span> <span class='kw'>&lt;-</span> <span class='fu'>splatSimulate</span>(<span class='no'>params</span>)
<span class='co'># Simulation with adjusted custom parameters</span>
<span class='no'>sim</span> <span class='kw'>&lt;-</span> <span class='fu'>splatSimulate</span>(<span class='no'>params</span>, <span class='kw'>mean.rate</span> <span class='kw'>=</span> <span class='fl'>0.6</span>, <span class='kw'>out.prob</span> <span class='kw'>=</span> <span class='fl'>0.2</span>)
<span class='co'># Simulate groups</span>
<span class='no'>sim</span> <span class='kw'>&lt;-</span> <span class='fu'>splatSimulate</span>(<span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"groups"</span>)
<span class='co'># Simulate paths</span>
<span class='no'>sim</span> <span class='kw'>&lt;-</span> <span class='fu'>splatSimulate</span>(<span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"paths"</span>)
<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#references">References</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Luke Zappia, Belinda Phipson, Alicia Oshlack.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

